pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java

```diff
@@ -269,31 +269,28 @@ public MessageImpl(String topic, String msgId, Map<String, String> properties,
         return msg;
     }
 
-    public static MessageImpl<byte[]> deserializeBrokerEntryMetaDataFirst(
-            ByteBuf headersAndPayloadWithBrokerEntryMetadata) throws IOException {
-        @SuppressWarnings("unchecked")
-        MessageImpl<byte[]> msg = (MessageImpl<byte[]>) RECYCLER.get();
-
-        msg.brokerEntryMetadata =
+    public static long getEntryTimestamp( ByteBuf headersAndPayloadWithBrokerEntryMetadata) throws IOException {
+        // get broker timestamp first if BrokerEntryMetadata is enabled with AppendBrokerTimestampMetadataInterceptor
+        BrokerEntryMetadata brokerEntryMetadata =
                 Commands.parseBrokerEntryMetadataIfExist(headersAndPayloadWithBrokerEntryMetadata);
-
-        if (msg.brokerEntryMetadata != null) {
-            msg.msgMetadata.clear();
-            msg.payload = null;
-            msg.messageId = null;
-            msg.topic = null;
-            msg.cnx = null;
-            msg.properties = Collections.emptyMap();
-            return msg;
+        if (brokerEntryMetadata != null && brokerEntryMetadata.hasBrokerTimestamp()) {
+            return brokerEntryMetadata.getBrokerTimestamp();
         }
-
+        // otherwise get the publish_time
+        MessageImpl<byte[]> msg = (MessageImpl<byte[]>) RECYCLER.get();
         Commands.parseMessageMetadata(headersAndPayloadWithBrokerEntryMetadata, msg.msgMetadata);
-        msg.payload = headersAndPayloadWithBrokerEntryMetadata;
-        msg.messageId = null;
-        msg.topic = null;
-        msg.cnx = null;
-        msg.properties = Collections.emptyMap();
-        return msg;
+        long entryTimestamp = msg.getPublishTime();
+        msg.recycle();
+        return entryTimestamp;
```

Comment:
It's better to use try-catch-finally block here like the previous code. `parseMessageMetadata` may throw `IllegalArgumentException` from LightProto though it's an unchecked exception.

Reply:
Do you mean `msg.getPublishTime()` may throws `IllegalArgumentException` if the `publish_time` is not set ?
Since `publish_time` is a  required field,  there must be something wroing if it's not set. 
Using try-catch-finally will hide the reason, so maybe, I think it's better to throws the exceptin.

Reply:
fixed